### ISSUE TITLE
Upgrading from version 11 -> 14

### ISSUE THREAD

**Issue by robbo10** [bitnami/bitnami-docker-postgresql#310]
**Description**

We are running the version 11 (11.10.0-debian-10-r52) and are in the process of upgrading to version 14.  This is on k8s where we have the data folder mounted through a PVC.

When attempting to run version 14 we get the following error:
```
The data directory was initialized by PostgreSQL version 11, which is not compatible with this version 14.0
```

We have managed to use `pg_dump` to export the data and load it onto a new PVC on the 14 image, however we would like to make use of `pg_upgrade` due to the size of the data.

When attempting to run `pg_upgrade` we get the following error:

```
could not look up effective user ID 1001: user does not exist
Failure, exiting
```

**Steps to reproduce the issue:**

1. Have a k8s pod running 11.10.0-debian-10-r52
2. Try to upgrade to version 14

**Describe the results you received:**

When trying to upgrade the image directly from 11 -> 14 we get:

```
The data directory was initialized by PostgreSQL version 11, which is not compatible with this version 14.0
```

When running a new pod which has the 14 image, we tried to mount both the updated PVC along with that of the 11 image to attempt a `pg_upgrade` however when running that we get:

```
could not look up effective user ID 1001: user does not exist
Failure, exiting
```

**Describe the results you expected:**

We are expecting to move the data from the version 11 image onto the container running version 14 within an appropriate timeframe, of which using `pg_dump` would not be efficient 

**Version**

This is running on kubernetes, EKS version 1.19

**Comment by Mauraza** [comment]
Hi @robbo10,

This container is non-root https://github.com/bitnami/bitnami-docker-postgresql#why-use-a-non-root-container to run commands of Postgres specify the user you configured by using the -U flag.

**Comment by robbo10** [comment]
@Mauraza  - This works when issuing the `pg_dump` command however i still get the error with `pg_upgrade`

```
I have no name!@pg-postgresql-14-0:/opt/bitnami/postgresql/bin$ pg_dump
pg_dump: error: local user with ID 1001 does not exist
I have no name!@pg-postgresql-14-0:/opt/bitnami/postgresql/bin$ pg_dump -U postgres
Password:

I have no name!@pg-postgresql-14-0:/opt/bitnami/postgresql/bin$ pg_upgrade

could not look up effective user ID 1001: user does not exist
Failure, exiting
I have no name!@pg-postgresql-14-0:/opt/bitnami/postgresql/bin$ pg_upgrade -U postgres

could not look up effective user ID 1001: user does not exist
Failure, exiting
```
Any ideas?

We did have to update the docker file to add custom packages.

Dockerfile is below
```
FROM docker.io/bitnami/postgresql:11.10.0-debian-10-r52

USER root
RUN apt-get update && apt-get install git make gcc vim postgresql-server-dev-11 -y \
    && git clone https://github.com/eulerto/wal2json.git && cd wal2json/ && make \
    && make install \
    && export PATH=/opt/bitnami/postgresql/bin:$PATH
USER postgresql
```

And for the container to run within k8s we needed the security context set to `1001`

**Comment by Mauraza** [comment]
Hi @robbo10,

Here you can find an example with Kubernetes and the chart https://github.com/bitnami/charts/issues/8025#issuecomment-964906018

If you use a container you can access as root you can use this command:
```bash
$ docker exec -ti -u root github_postgresql_1 bash
```
and then follow the same instruction that appears in the [post](https://github.com/bitnami/charts/issues/8025#issuecomment-964906018)
```bash
$ kubectl exec -it postgresql-postgresql-0 -- bash
root@postgresql-postgresql-0:/# postgres --version
postgres (PostgreSQL) 12.8
root@postgresql-postgresql-0:/# . /opt/bitnami/scripts/libos.sh
root@postgresql-postgresql-0:/# ensure_group_exists postgres -i 1001
root@postgresql-postgresql-0:/# ensure_user_exists postgres -i 1001 -g postgres
root@postgresql-postgresql-0:/tmp# mv /bitnami/postgresql/data /bitnami/postgresql/olddata
root@postgresql-postgresql-0:/tmp# mkdir -p /bitnami/postgresql/data mkdir -p /bitnami/postgresql/oldbin
root@postgresql-postgresql-0:/tmp# chown -R postgres:postgres /bitnami/postgresql/data /bitnami/postgresql/olddata
root@postgresql-postgresql-0:/# cd /tmp/
root@postgresql-postgresql-0:/# curl --remote-name --silent https://downloads.bitnami.com/files/stacksmith/postgresql-11.13.0-14-linux-amd64-debian-10.tar.gz
root@postgresql-postgresql-0:/# tar --extract --directory /bitnami/postgresql/oldbin/ --file postgresql-11.13.0-14-linux-amd64-debian-10.tar.gz --strip-components=4 postgresql-11.13.0-14-linux-amd64-debian-10/files/postgresql/bin
root@postgresql-postgresql-0:/tmp# gosu postgres initdb -E UTF8 -D /bitnami/postgresql/data -U postgres
(...)
root@postgresql-postgresql-0:/tmp# gosu postgres pg_upgrade -c -b /bitnami/postgresql/oldbin -B /opt/bitnami/postgresql/bin -d /bitnami/postgresql/olddata -D /bitnami/postgresql/data
(...)
```

**Comment by robbo10** [comment]
@Mauraza - awesome thanks so much will look into this :)

**Comment by qdupuy** [comment]
@robbo10 

I'm in a Kubernetes environment.

However, I couldn't do what he said due to time constraints, I had to deliver to my client quickly. So I created a postgres 13 instance next to it and did a pg_dumpall to integrate it with the new instance

**Comment by Mauraza** [comment]
Hi @robbo10,

Has it worked?

**Comment by robbo10** [comment]
@Mauraza  - I wasn't able to get it working we are not using helm for the install, just applying the yaml through kubectl so I couldn't set  `diagnosticMode.enabled=true` .  Each time I tried to run the new image the deployment keeps failing with 

```
The data directory was initialized by PostgreSQL version 11, which is not compatible with this version 14.0
```

From looking at this reply I don't think `pg_upgrade` is possible in production k8s environments? https://github.com/bitnami/charts/issues/8025#issuecomment-968601608

**Comment by Mauraza** [comment]
Hi @robbo10,

if you don't have `diagnosticMode.enabled=true` you need to access the pod with permission of root and then run with the same commands.

**Comment by javsalgar** [comment]
Hi,

As there are a lot of interactions in this issue. Could you add more details on the exact issue you are facing and the steps to reproduce it?

**Comment by jnt2007** [comment]
Hi @Mauraza 
TL;DR no, it is not working

Checked [steps](https://github.com/bitnami/bitnami-docker-postgresql/issues/310#issuecomment-966976434) for my case 11.14.0-debian-10-r17 >>> 14.3.0-debian-10-r17
Last step fails because new postgres libs not working with old binaries. Details below
```
root@postgres-cm-0:/bitnami/postgresql# gosu postgres pg_upgrade --old-bindir /bitnami/postgresql/oldbin --new-bindir /opt/bitnami/postgresql/bin --old-datadir /bitnami/postgresql/olddata --new-datadir /bitnami/postgresql/data 
Performing Consistency Checks
-----------------------------
Checking cluster versions                                   ok

*failure*
Consult the last few lines of "pg_upgrade_server.log" for
the probable cause of the failure.

connection to server on socket "/bitnami/postgresql/.s.PGSQL.50432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?

could not connect to source postmaster started with the command:
"/bitnami/postgresql/oldbin/pg_ctl" -w -l "pg_upgrade_server.log" -D "/bitnami/postgresql/olddata" -o "-p 50432 -b  -c listen_addresses='' -c unix_socket_permissions=0700 -c unix_socket_directories='/bitnami/postgresql'" start
Failure, exiting
```
```
root@postgres-cm-0:/bitnami/postgresql# tail -f pg_upgrade_server.log
-----------------------------------------------------------------

command: "/bitnami/postgresql/oldbin/pg_ctl" -w -l "pg_upgrade_server.log" -D "/bitnami/postgresql/olddata" -o "-p 50432 -b  -c listen_addresses='' -c unix_socket_permissions=0700 -c unix_socket_directories='/bitnami/postgresql'" start >> "pg_upgrade_server.log" 2>&1
waiting for server to start....2022-07-13 17:54:10.518 GMT [523] FATAL:  could not load library "/opt/bitnami/postgresql/lib/pgaudit.so": /opt/bitnami/postgresql/lib/pgaudit.so: undefined symbol: GetCommandTagName
2022-07-13 17:54:10.518 GMT [523] LOG:  database system is shut down
 stopped waiting
pg_ctl: could not start server
Examine the log output.
```

**Comment by javsalgar** [comment]
It seems that it is using the symbols from the incorrect libraries. You may need to tweak the LD_LIBRARY_PATH variable to ensure it loads the proper ones.

**Comment by jnt2007** [comment]
Hi @javsalgar 
Thank you for your attention.
I tried this env variable, but it does not affect to `pg_upgrade` utility
```
root@postgres-cm-0:/bitnami/postgresql# export LD_LIBRARY_PATH=/bitnami/postgresql/oldlib/
root@postgres-cm-0:/bitnami/postgresql# 
root@postgres-cm-0:/bitnami/postgresql# gosu postgres pg_upgrade --old-bindir /bitnami/postgresql/oldbin --new-bindir /opt/bitnami/postgresql/bin --old-datadir /bitnami/postgresql/olddata --new-datadir /bitnami/postgresql/data 
Performing Consistency Checks
-----------------------------
Checking cluster versions                                   ok

*failure*
Consult the last few lines of "pg_upgrade_server.log" for
the probable cause of the failure.

connection to server on socket "/bitnami/postgresql/.s.PGSQL.50432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?

could not connect to source postmaster started with the command:
"/bitnami/postgresql/oldbin/pg_ctl" -w -l "pg_upgrade_server.log" -D "/bitnami/postgresql/olddata" -o "-p 50432 -b  -c listen_addresses='' -c unix_socket_permissions=0700 -c unix_socket_directories='/bitnami/postgresql'" start
Failure, exiting
root@postgres-cm-0:/bitnami/postgresql# tail -f pg_upgrade_server.log
-----------------------------------------------------------------

command: "/bitnami/postgresql/oldbin/pg_ctl" -w -l "pg_upgrade_server.log" -D "/bitnami/postgresql/olddata" -o "-p 50432 -b  -c listen_addresses='' -c unix_socket_permissions=0700 -c unix_socket_directories='/bitnami/postgresql'" start >> "pg_upgrade_server.log" 2>&1
waiting for server to start....2022-07-13 17:54:10.518 GMT [523] FATAL:  could not load library "/opt/bitnami/postgresql/lib/pgaudit.so": /opt/bitnami/postgresql/lib/pgaudit.so: undefined symbol: GetCommandTagName
2022-07-13 17:54:10.518 GMT [523] LOG:  database system is shut down
 stopped waiting
pg_ctl: could not start server
Examine the log output.
```

Looks like `pg_upgrade` ignores such option and I have theory why.
`pg_upgrade` trying to run both postgres instances at the same time and it doesn't know for which instance apply this env var.
I mean if some var is real for `pg_upgrade` it should be something like OLD_LD_LIBRARY_PATH and NEW_LD_LIBRARY_PATH.

**Comment by javsalgar** [comment]
Hi,

In this case I'm not sure how the issue would be fixed. Maybe the PostgreSQL devs can provide more insight on how to proceed?
